### PR TITLE
Add `carton init` command

### PR DESCRIPTION
**Pull request by carson-katri**
This adds basic `carton init` support:

```
carton init HelloWorld
```

It also supports templates:
```
carton init --list-templates
carton init HelloWorld --template tokamak
```

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
This is brilliant, thanks a ton! I'll have a closer look tomorrow.

File: Sources/carton/Model/Template.swift [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                        terminal)
+      try createManifest(fileSystem: fileSystem,
+                         project: project,
+                         terminal)
+    }
+  }
+}
+
+extension Templates {
+  struct Tokamak : Template {
```
Comment (MaxDesiatov) [inline Sources/carton/Model/Template.swift:136]: **nitpick**: would you be able to run `swiftformat .` in the project or pre-commit installed as described in [the "Coding Style" section](https://github.com/swiftwasm/carton#coding-style) of `README.md`?

```suggestion
  struct Tokamak: Template {
```

File: Sources/carton/Model/Template.swift [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        let renderer = DOMRenderer(ContentView(), div)
+        _ = body.appendChild!(div)
+        """
+          .write(to: $0)
+      }
+      try fileSystem.writeFileContents(project.path.appending(components: "Sources", project.name, "ContentView.swift")) {
+        """
+        import TokamakDOM
+
+        struct ContentView : View {
```
Comment (MaxDesiatov) [inline Sources/carton/Model/Template.swift:175]: **nitpick:** I'm probably fine with the default 4 spaces indentation for new projects, but not sure if spaces before colons is consistent with the colon formatting in say `var body: some View`. Up to you, this is not blocking at all 🙂

```suggestion
        struct ContentView: View {
```

**Comment by carson-katri** [comment]
It shouldn't require that option since it's optional:
```swift
var template: Templates?
```
I changed it to `Optional` in https://github.com/swiftwasm/carton/pull/54/commits/2af3f8444ed19ae3c69fa3a878bb162958890d36

**Comment by MaxDesiatov** [comment]
Apologies, I forgot to rebuild. Works for me now.

File: Sources/carton/Commands/Init.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    else { fatalError("failed to create an instance of `TerminalController`") }
+
+    if listTemplates {
+      Templates.allCases.forEach {
+        terminal.write($0.rawValue, inColor: .green, bold: true)
+        terminal.write("\t\($0.template.description)\n")
+      }
+    } else {
+      guard let name = name else {
+        terminal.write("No name specified.\n", inColor: .red)
```
Comment (MaxDesiatov) [inline Sources/carton/Commands/Init.swift:47]: I wonder how hard would it be to infer the name from the name of the current directory, similarly to what `swift package init` does?

Comment (carson-katri) [reply Sources/carton/Commands/Init.swift:47]: Probably easily with `localFileSystem.currentWorkingDirectory?.dirname`, then convert to PascalCase if necessary.

Comment (carson-katri) [reply Sources/carton/Commands/Init.swift]: I've reworked the command to support no name input. This also allowed me to make `list-templates` a subcommand.

File: Sources/carton/Commands/Init.swift [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      terminal.write("Failed to get current working directory.\n", inColor: .red)
+      return
+    }
+    let template = self.template ?? .basic
+    terminal.write("Creating new project with template ")
+    terminal.write("\(template.rawValue)", inColor: .green)
+    terminal.write(" in ")
+    terminal.write("\(name)\n", inColor: .cyan)
+
+    guard let packagePath = self.name == nil ? localFileSystem.currentWorkingDirectory : AbsolutePath(name, relativeTo: currentDir) else {
```
Comment (MaxDesiatov) [inline Sources/carton/Commands/Init.swift:52]: Sorry, `danger` is not running on the fork PR, another line length nitpick

```suggestion
    guard let packagePath = self.name == nil ? 
      localFileSystem.currentWorkingDirectory : 
      AbsolutePath(name, relativeTo: currentDir) 
    else {
```

File: Sources/carton/Commands/Init.swift [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      terminal.write("Path to project could be created.\n", inColor: .red)
+      return
+    }
+    try localFileSystem.createDirectory(packagePath)
+    try template.template.create(on: localFileSystem,
+                                 project: .init(name: name, path: packagePath, inPlace: self.name == nil),
+                                 terminal)
```
Comment (MaxDesiatov) [inline Sources/carton/Commands/Init.swift:59]:
```suggestion
    try template.template.create(
      on: localFileSystem,
      project: .init(name: name, path: packagePath, inPlace: self.name == nil),
      terminal
    )
```

File: Sources/carton/Model/Template.swift [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                        terminal)
+      try createManifest(fileSystem: fileSystem,
+                         project: project,
+                         dependencies: [
+                           .init(name: "Tokamak", url: "https://github.com/swiftwasm/Tokamak", version: .branch("main")),
+                         ],
+                         targetDepencencies: [
+                           .init(name: "TokamakDOM", package: "Tokamak"),
+                         ],
+                         terminal)
```
Comment (MaxDesiatov) [inline Sources/carton/Model/Template.swift:157]:
```suggestion
      try createManifest(
        fileSystem: fileSystem,
        project: project,
        dependencies: [
          .init(name: "Tokamak", url: "https://github.com/swiftwasm/Tokamak", version: .branch("main")),
        ],
        targetDepencencies: [
          .init(name: "TokamakDOM", package: "Tokamak"),
        ],
        terminal
      )
```

File: Sources/carton/Model/Template.swift [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      try fileSystem.writeFileContents(project.path.appending(components: "Sources", project.name, "main.swift")) {
```
Comment (MaxDesiatov) [inline Sources/carton/Model/Template.swift:160]:
```suggestion
      try fileSystem.writeFileContents(project.path.appending(
        components: "Sources", 
        project.name, 
        "main.swift"
      )) {
```

File: Sources/carton/Model/Template.swift [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+          platforms: [.macOS(.v10_15)],
+          products: [
+              .executable(name: "\(project.name)", targets: ["\(project.name)"])
+          ],
+          dependencies: [
+              \(dependencies.map(\.description).joined(separator: ",\n"))
+          ],
+          targets: [
+              .target(
+                  name: "HelloWorld",
```
Comment (MaxDesiatov) [inline Sources/carton/Model/Template.swift:101]:
```suggestion
                  name: "\(project.name)",
```

Comment (MaxDesiatov) [reply Sources/carton/Model/Template.swift:101]: The default `Package.swift` doesn't work for me without this

Comment (carson-katri) [reply Sources/carton/Model/Template.swift:101]: Oo good catch

**Comment by carson-katri** [comment]
In the future it may be nice to move the templates into git repos, like vapor does

**Pull request merged**